### PR TITLE
Add Punctures docs and system

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

First step to solve for puncture initial data.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Elliptic/Systems/Punctures/Test_Equations.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+void test_equations(const DataVector& used_for_size) {
+  const double eps = 1.e-12;
+  const auto seed = std::random_device{}();
+  const double fill_result_tensors = 0.;
+  pypp::check_with_random_values<3>(
+      &Punctures::add_sources, "Equations", {"sources"},
+      {{{-1., 1.}, {-1., 1.}, {-1., 1.}}}, used_for_size, eps, seed,
+      fill_result_tensors);
+  pypp::check_with_random_values<4>(
+      &Punctures::add_linearized_sources, "Equations", {"linearized_sources"},
```
Comment (nilsdeppe) [inline tests/Unit/Elliptic/Systems/Punctures/Test_Equations.cpp:30]: Could these test `Sources` and `LinearizedSources`?

File: src/Elliptic/Systems/Punctures/BoundaryConditions/Flatness.hpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  Flatness& operator=(Flatness&&) = default;
+  ~Flatness() = default;
+
+  /// \cond
+  explicit Flatness(CkMigrateMessage* m) : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Flatness);
+  /// \endcond
+
+  std::unique_ptr<domain::BoundaryConditions::BoundaryCondition> get_clone()
```
Comment (nilsdeppe) [inline src/Elliptic/Systems/Punctures/BoundaryConditions/Flatness.hpp:46]: add to test?

File: tests/Unit/Elliptic/Systems/Punctures/BoundaryConditions/Test_Flatness.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  elliptic::apply_boundary_condition<Linearized, void, tmpl::list<Flatness>>(
+      boundary_condition, box, Direction<3>::upper_xi(), make_not_null(&field),
+      n_dot_field_gradient);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Punctures.BoundaryConditions.Flatness",
+                  "[Unit][Elliptic]") {
+  // Test factory-creation
+  const auto created = TestHelpers::test_factory_creation<
```
Comment (nilsdeppe) [inline tests/Unit/Elliptic/Systems/Punctures/BoundaryConditions/Test_Flatness.cpp]: `serialize_and_deserialize(...).get_clone()`

File: src/Elliptic/Systems/Punctures/Sources.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+void Sources::apply(const gsl::not_null<Scalar<DataVector>*> puncture_equation,
+                    const Scalar<DataVector>& alpha,
+                    const Scalar<DataVector>& beta,
+                    const Scalar<DataVector>& field,
+                    const tnsr::I<DataVector, 3>& /*field_flux*/) {
+  add_sources(puncture_equation, alpha, beta, field);
+}
+
+void Sources::apply(
```
Comment (nilsdeppe) [inline src/Elliptic/Systems/Punctures/Sources.cpp:40]: lots of untested code here...

File: src/Elliptic/Systems/Punctures/BoundaryConditions/Flatness.hpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      const override {
+    return std::make_unique<Flatness>(*this);
+  }
+
+  std::vector<elliptic::BoundaryConditionType> boundary_condition_types()
+      const override {
+    return {1, elliptic::BoundaryConditionType::Neumann};
```
Comment (nilsdeppe) [inline src/Elliptic/Systems/Punctures/BoundaryConditions/Flatness.hpp]: either switch to `()` or `{elliptic::BoundaryConditionType::Neumann}`. The current is a bit tough to parse with `std::vector`s annoying constructors. I think one (or both) of the suggestions should work and are unambiguous/avoid the silliness that is std::vector construction

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Elliptic/Systems/Punctures/Test_Equations.cpp (hunk 1 above)
Comment (nilsvu) [reply tests/Unit/Elliptic/Systems/Punctures/Test_Equations.cpp:30]: added test below

File: src/Elliptic/Systems/Punctures/Sources.cpp (hunk 4 above)
Comment (nilsvu) [reply src/Elliptic/Systems/Punctures/Sources.cpp:40]: fixed in another commit

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**